fix(request-response): don't keep duplicate addresses

b-zee opened the pull request:
## Description

Resolves: #4699.

<!--
Please write a summary of your changes and why you made them.
This section will appear as the commit message after merging.
Please craft it accordingly.
For a quick primer on good commit mesages, check out this blog post: https://cbea.ms/git-commit/

Please include any relevant issues in here, for example:

Related https://github.com/libp2p/rust-libp2p/issues/ABCD.
Fixes https://github.com/libp2p/rust-libp2p/issues/XYZ.
-->

## Notes & open questions

In `kad`, there is a special `Addresses` struct that might be worth reusing for this functionality.

I will make further CHANGELOG edits etc if this fix is thought to be sensible.
<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

b-zee committed:
fix: deduplicate autonat add_server addresses

fixes #4699

joshuef commented on protocols/request-response/src/lib.rs:
```diff
@@ -437,8 +437,19 @@ where
+    ///
+    /// Returns true if the address was added, false otherwise (i.e. if the
+    /// address is already in the list).
+    pub fn add_address(&mut self, peer: &PeerId, address: Multiaddr) -> bool {
```
I wonder if its work making this take `&Multiaddr`and then only cloning if we need to add it.

It'd mean the consumer of the API doesn't have to take ownership to use it (which may involve a pointless clone, as we have atm).

mxinden replied:
Note that `Multiaddr` is a wrapper around an `Arc<Vec<_>>`. Thus I doubt additional clones have a performance impact here.

https://github.com/multiformats/rust-multiaddr/blob/d0e9ec219645ce8bd8ae7926e4b7316a67c35f84/src/lib.rs#L40-L45

joshuef replied:
Ah nice one. Good to know! 🙇

mxinden reviewed (commented):
How about using a `HashSet` instead?

``` diff
modified   protocols/request-response/src/lib.rs
@@ -337,7 +337,7 @@ where
     /// reachable addresses, if any.
     connected: HashMap<PeerId, SmallVec<[Connection; 2]>>,
     /// Externally managed addresses via `add_address` and `remove_address`.
-    addresses: HashMap<PeerId, SmallVec<[Multiaddr; 6]>>,
+    addresses: HashMap<PeerId, HashSet<Multiaddr>>,
     /// Requests that have not yet been sent and are waiting for a connection
     /// to be established.
     pending_outbound_requests: HashMap<PeerId, SmallVec<[RequestProtocol<TCodec>; 10]>>,
```

b-zee committed:
keep peers in HashSet instead of SmallVec

This prevents us from adding duplicates into the list of peer addresses we have.

b-zee committed:
update request-response CHANGELOG.md for #4700

b-zee committed:
change (&HashSet)::into_iter to equivalent iter

thomaseizinger committed:
Merge branch 'master' into fix-autonat-add-server-deduplicate

mergify[bot] committed:
Merge branch 'master' into fix-autonat-add-server-deduplicate
